File: 2024/Day3/main.dart
import 'dart:convert';
import 'dart:io';

int solve1(String input) {
  var total = 0;

  final pattern = RegExp(r'mul\((\d{1,3}),(\d{1,3})\)');
  final matches = pattern.allMatches(input);

  for (final match in matches) {
    if (match.group(1) != null && match.group(2) != null) {
      final x = int.parse(match.group(1)!);
      final y = int.parse(match.group(2)!);
      total += x * y;
    }
  }

  return total;
}

int solve2(String input) {
  var total = 0;
  var enabled = true;
  final pattern = RegExp(r"(do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\))");
  final matches = pattern.allMatches(input);

  for (final match in matches) {
    final instruction = match.group(1);
    if (instruction == null) continue;

    if (instruction == 'do()') {
      enabled = true;
    } else if (instruction == "don't()") {
      enabled = false;
    } else if (enabled && instruction.startsWith('mul')) {
      if (match.group(2) != null && match.group(3) != null) {
        final x = int.parse(match.group(2)!);
        final y = int.parse(match.group(3)!);
        total += x * y;
      }
    }
  }

  return total;
}

Future<void> main() async {
  String input = await stdin.transform(utf8.decoder).join();

  print('Part 1: ${solve1(input)}');
  print('Part 2: ${solve2(input)}');
}

File: 2024/Day3/main.py
import re
import sys

def solve1(input: str) -> int:
    total = 0

    pattern = re.compile(r'mul\((\d{1,3}),(\d{1,3})\)')

    for match in pattern.finditer(input):
        x = int(match.group(1))
        y = int(match.group(2))
        total += x * y

    return total

def solve2(input: str) -> int:
    total = 0
    enabled = True

    pattern = re.compile(r"(do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\))")

    for match in pattern.finditer(input):
        instruction = match.group(1)

        if instruction == 'do()':
            enabled = True
        elif instruction == "don't()":
            enabled = False
        elif enabled and instruction.startswith('mul'):
            x = int(match.group(2))
            y = int(match.group(3))
            total += x * y

    return total

data = sys.stdin.read()

print(f'Part 1: {solve1(data)}')
print(f'Part 2: {solve2(data)}')
